### PR TITLE
sets, dicts, and sorted arrays in the IR (take 2)

### PR DESCRIPTION
**Pull request by catoverdrive**
builds on #3426, will rebase once that goes in. (rebased)

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/ArraySorter.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        (i < array.size && i.cne(n + 1)).mux(
+          Code(
+            array.setMissing(n + 1, array.isMissing(i)),
+            array.isMissing(n + 1).mux(
+              Code._empty,
+              array.update(n + 1, array(i)))),
+          Code._empty),
+        n += 1),
+      array.setSize(n))
+  }
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/ArraySorter.scala]: i added a `sortIntoRegion(distinct: Boolean): Code[Long]` which condenses the three calls. I think this contains all of the emit logic that the ArraySorter is actually in charge of.

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 2]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
     }
     p.check()
   }
+
+  @Test def testSortOnRandomArray() {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: add one test that sorts an NA: Array(Tint32())

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 3]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+    val compareGen = Type.genArb.flatMap(t => Gen.zip(Gen.const(t), TArray(t).genNonmissingValue))
+    val p = Prop.forAll(compareGen) { case (t, a) =>
+      val ir = ArraySort(GetTupleElement(In(0, TTuple(TArray(t))), 0))
+      val fb = EmitFunctionBuilder[Region, Long, Boolean, Long]
+
+      Emit(ir, fb)
+
+      val f = fb.result()()
+
+      val region = Region()
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: use `Region.scoped`

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 4]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+    val p = Prop.forAll(compareGen) { case (t, a) =>
+      val array = a.asInstanceOf[IndexedSeq[Any]] ++ a.asInstanceOf[IndexedSeq[Any]]
+      val ir = ToSet(GetTupleElement(In(0, TTuple(TArray(t))), 0))
+      val fb = EmitFunctionBuilder[Region, Long, Boolean, Long]
+
+      Emit(ir, fb)
+
+      val f = fb.result()()
+
+      val region = Region()
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: Region.scoped

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 5]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+    val p = Prop.forAll(compareGen) { case (telt, a) =>
+      val array: IndexedSeq[Row] = a.asInstanceOf[IndexedSeq[Row]] ++ a.asInstanceOf[IndexedSeq[Row]]
+      val ir = ToDict(GetTupleElement(In(0, TTuple(TArray(telt))), 0))
+      val fb = EmitFunctionBuilder[Region, Long, Boolean, Long]
+
+      Emit(ir, fb)
+
+      val f = fb.result()()
+
+      val region = Region()
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: region.scoped

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 6]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+      val res = f(region, off, false)
+      val actual = SafeIndexedSeq(TArray(t), region, res)
+      val expected = a.asInstanceOf[IndexedSeq[Any]].sorted(t.ordering.toOrdering)
+
+      expected == actual
+    }
+    p.check()
+  }
+
+  @Test def testToSetOnRandomDuplicatedArray() {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: test on NA array

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 7]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+      val res = f(region, off, false)
+      val actual = SafeIndexedSeq(TArray(t), region, res)
+      val expected = a.asInstanceOf[IndexedSeq[Any]].sorted(t.ordering.toOrdering).distinct
+
+      expected == actual
+    }
+    p.check()
+  }
+
+  @Test def testToDictOnRandomDuplicatedArray() {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: test on NA array

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 8]
```diff
@@ -90,4 +90,94 @@ class OrderingSuite {
...
+    val compareGen = Gen.zip(Type.genArb, Type.genArb).flatMap {
+      case (k, v) => Gen.zip(Gen.const(TTuple(k, v)), TArray(TTuple(k, v)).genNonmissingValue)
+    }
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: sometimes this is easier to read as
```scala
val compareGen = for {
  kt <- Type.genArb
  vt <- Type.genArb
  dictArrayType = TArray(TTuple(kt, vt))
  kv <- dictArrayType.genNonMissingValue
} yield (dictArrayType, kv)
```

**Comment by danking** [comment]
Sorry my bad, will look today.

File: src/test/scala/is/hail/expr/ir/OrderingSuite.scala [hunk 9]
```diff
@@ -41,53 +41,182 @@ class OrderingSuite {
   }
 
   @Test def testRandomOpsAgainstExtended() {
-    val compareGen = Type.genArb.flatMap(t => Gen.zip(Gen.const(t), t.genNonmissingValue, t.genNonmissingValue))
+    val compareGen = for {
+      t <- Type.genArb
+      a1 = t.genNonmissingValue
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/OrderingSuite.scala]: should be a right arrow

**Pull request merged**